File: dart_backend/lib/middleware/audit_logger.dart
// lib/middleware/audit_logger.dart
// Non-blocking audit trail — port of audit.logger.js
// Writes to audit_logs table AFTER the route responds (non-blocking).

import 'dart:async';
import 'dart:convert';
import 'package:shelf/shelf.dart';
import '../config/supabase_service.dart';
import '../models/user_session.dart';

/// auditLog(action) — records sensitive operations to audit_logs table.
/// If DB write fails, logs to console but never crashes the request.
/// BUG-10 FIX: Always place auditLog BEFORE sensitiveLimiter in chain.
Middleware auditLog(String action) {
  return (Handler inner) {
    return (Request request) async {
      final session = request.context['session'] as UserSession?;

      // Read body once for audit (re-inject for handler)
      final bodyStr = await request.readAsString();
      final updatedRequest = request.change(body: bodyStr);

      // Run the actual handler first
      final response = await inner(updatedRequest);

      // Non-blocking audit write
      unawaited(_writeAudit(
        action: action,
        session: session,
        request: request,
        bodyStr: bodyStr,
        responseStatus: response.statusCode,
      ));

      return response;
    };
  };
}

Future<void> _writeAudit({
  required String action,
  required UserSession? session,
  required Request request,
  required String bodyStr,
  required int responseStatus,
}) async {
  try {
    final sanitized = _sanitizeBody(bodyStr);
    final payloadSummary = jsonEncode({
      'params': request.url.queryParameters,
      'body': sanitized,
    });

    await SupabaseService.insert('audit_logs', {
      'action': action,
      'profile_id': session?.profileId,
      'role': session?.role,
      'branch_id': session?.queryBranchId ?? session?.branchId,
      'ip_address': _getIp(request),
      'user_agent': request.headers['user-agent'],
      'request_path': '/${request.url.path}',
      'request_method': request.method,
      'response_status': responseStatus,
      'payload_summary': payloadSummary.length > 500
          ? payloadSummary.substring(0, 500)
          : payloadSummary,
      'created_at': DateTime.now().toUtc().toIso8601String(),
    });
  } catch (e) {
    // Audit failure must NEVER break the app
    print('⚠️  Audit log write failed: $e');
  }
}

Map<String, dynamic> _sanitizeBody(String bodyStr) {
  if (bodyStr.isEmpty) return {};
  try {
    final body = jsonDecode(bodyStr) as Map<String, dynamic>;
    final safe = Map<String, dynamic>.from(body);
    // Never log passwords or tokens
    safe.remove('password');
    safe.remove('password_hash');
    safe.remove('access_token');
    safe.remove('refresh_token');
    return safe;
  } catch (_) {
    return {};
  }
}

String _getIp(Request req) =>
    req.headers['x-forwarded-for']?.split(',').first.trim() ??
    req.headers['x-real-ip'] ??
    'unknown';

File: old -backend/src/middleware/audit.logger.js
/**
 * AUDIT LOGGER MIDDLEWARE
 * Records every sensitive action to the `audit_logs` table in Supabase.
 * Provides a full trail: who did what, when, from which IP.
 *
 * Usage: router.patch('/approve', requireAuth, auditLog('APPROVE_MARKSHEET'), handler)
 */
const { supabase } = require('../config/supabase');

/**
 * auditLog(action)
 * Logs to `audit_logs` table AFTER the route handler responds.
 * If DB write fails, we log to console but don't fail the request.
 */
const auditLog = (action) => async (req, res, next) => {
  // Capture original json method to intercept response status
  const originalJson = res.json.bind(res);
  let responseStatus = 200;

  res.json = function (body) {
    responseStatus = res.statusCode;
    return originalJson(body);
  };

  // Continue to handler first
  next();

  // After handler, write audit record asynchronously (non-blocking)
  setImmediate(async () => {
    try {
      await supabase.from('audit_logs').insert({
        action,
        profile_id:   req.profileId  || null,
        role:         req.role        || null,
        branch_id:    req.queryBranchId || req.branchId || null,
        ip_address:   req.ip,
        user_agent:   req.headers['user-agent'] || null,
        request_path: req.originalUrl,
        request_method: req.method,
        response_status: responseStatus,
        payload_summary: JSON.stringify({
          params: req.params,
          // Never log passwords or tokens
          body: sanitizeBody(req.body),
        }).substring(0, 500),
        created_at: new Date().toISOString(),
      });
    } catch (err) {
      // Audit failure must NEVER break the app — just log
      console.error('⚠️  Audit log write failed:', err.message);
    }
  });
};

function sanitizeBody(body) {
  if (!body) return {};
  const safe = { ...body };
  delete safe.password;
  delete safe.password_hash;
  delete safe.access_token;
  delete safe.refresh_token;
  return safe;
}

module.exports = { auditLog };
